rules.go

```diff
@@ -0,0 +1,217 @@
+)
+
+// Structs
+
+type Request struct {
+	// parsed full URL of the request
+	URL *url.URL
+	// a value of Origin header
+	Origin string
+	// a value of Referer header
+	Referer string
+	// Defines is request looks like XHLHttpRequest
+	IsXHR bool
+}
+
+type Rule struct {
+	raw         string
+	ruleText    string
+	regexString string
+	options     map[string]bool
+	isException bool
+	domains     map[string]bool
+}
+
+// Interfaces
+
+type RuleSet interface {
+	Allow(*Request) bool
+}
+
+// Allow methods
+
+func (rule *Rule) Allow(url string) bool {
+	// TODO
+	return false
+}
+
+//
+
+func ParseRule(ruleText string) (*Rule, error) {
+	rule := &Rule{
+		domains: map[string]bool{},
+		options: map[string]bool{},
+	}
+	rule.raw = ruleText
+	rule.ruleText = strings.TrimSpace(ruleText)
+
+	isComment := strings.Contains(rule.ruleText, "!") || strings.Contains(rule.ruleText, "[Adblock")
+	if isComment {
+		return nil, ErrSkipComment
+	}
+
+	isHTMLRule := strings.Contains(rule.ruleText, "##") || strings.Contains(rule.ruleText, "#@#")
+	if isHTMLRule {
+		return nil, ErrSkipHTML
+	}
+
+	rule.isException = strings.HasPrefix(rule.ruleText, "@@")
+	if rule.isException {
+		rule.ruleText = rule.ruleText[2:]
+	}
+
+	rule.options = make(map[string]bool)
+	if strings.Contains(rule.ruleText, "$") {
+		var option string
+
+		parts := strings.SplitN(rule.ruleText, "$", 2)
+		length := len(parts)
+
+		if length > 0 {
+			rule.ruleText = parts[0]
+		}
+
+		if length > 1 {
+			option = parts[1]
+		}
+
+		options := strings.Split(option, ",")
+		for _, option := range options {
+			if strings.HasPrefix(option, "domain=") {
+				rule.domains = parseDomainOption(option)
+			} else {
+				rule.options[strings.TrimPrefix(option, "~")] = !strings.HasPrefix(option, "~")
+			}
+		}
+	} else {
+		rule.domains = make(map[string]bool)
+	}
+
+	rule.regexString = ruleToRegexp(rule.ruleText)
+
+	return rule, nil
+}
+
+func NewRuleSet(rules []*Rule) (RuleSet, error) {
+	return nil, nil
+}
+
+func (rule *Rule) OptionsKeys() []string {
+	opts := []string{}
+	for option := range rule.options {
+		opts = append(opts, option)
+	}
+
+	if rule.domains != nil && len(rule.domains) > 0 {
+		opts = append(opts, "domain")
+	}
+
+	return opts
+}
+
+func parseDomainOption(text string) map[string]bool {
+	domains := text[len("domain="):]
+	parts := strings.Split(domains, "|")
+	opts := make(map[string]bool, len(parts))
+
+	for _, part := range parts {
+		opts[strings.TrimPrefix(part, "~")] = !strings.HasPrefix(part, "~")
+	}
+
+	return opts
+}
+
+func ruleToRegexp(text string) string {
+	// Convert AdBlock rule to a regular expression.
+	if text == "" {
+		return ".*"
+	}
+
+	// Check if the rule isn't already regexp
+	length := len(text)
+	if length >= 2 && text[:1] == "/" && text[length-1:] == "/" {
+		return text[1 : length-1]
+	}
+
+	// escape special regex characters
+	rule := escapeSpecialRegxp.ReplaceAllStringFunc(text, regexp.QuoteMeta)
```

Comment:
`regexp.QuoteMeta` already escapes special characters. This is really not required. `escapeSpecialRegexp` has to be deleted